C++/C.sublime-syntax

```diff
@@ -13,1427 +14,735 @@ first_line_match: |-
+      captures:
+        1: keyword.operator.c
+        2: entity.name.type.typedef.c
+        3: punctuation.section.group.begin.c
+      set: function.params
+
+    - match: ({{identifier}})\s*(?=;)
+      captures:
+        1: entity.name.type.typedef.c
+      pop: 1
+    - match: ; # Fallback
+      pop: 1
+    - include: expression
+  type.enum:
+    - meta_scope: meta.enum.c
+    - include: expression.call
+    - match: ({{identifier}})\s*(?=;)
       captures:
-        1: punctuation.definition.comment.c
+        1: entity.name.enum.forward-decl.c
       pop: 1
-    - match: (\\)$\n?
+    - match: (?:{{identifier}}\s*)*?({{identifier}})?\s*(?:{{likely_macro}}\s*)*(?=\{|$)
       captures:
-        1: punctuation.separator.continuation.c
+        1: entity.name.enum.c
+      set:
+        - match: \{
+          scope: meta.enum.c punctuation.section.block.begin.c
+          set: type.enum-body
+        - match: (?=\S)
+          fail: enum
+    - match: (?=\S)
+      fail: enum
+  type.enum-body:
+    - meta_content_scope: meta.block.c
+    - match: \b{{identifier}}\b
+      scope: entity.name.constant.c
+      push:
+        - match: (?=,|})
+          pop: 1
+        - include: expression
+    - include: punctuation.comma
+    - match: \}
+      scope: punctuation.section.block.end.c
+      pop: 1
+  type.struct:
+    - meta_scope: meta.struct.c
+    - include: expression.call
+    - match: ({{identifier}})\s*(?=;)
+      captures:
+        1: entity.name.enum.forward-decl.c
+      pop: 1
+    - match: (?:{{identifier}}\s*)*?({{identifier}})?\s*(?:{{likely_macro}}\s*)*(?=\{|$)
+      captures:
+        1: entity.name.struct.c
+      set:
+        - match: \{
+          scope: meta.struct.c punctuation.section.block.begin.c
+          set: enclosure.block
+        - match: (?=\S)
+          fail: struct
+    - match: (?=\S)
+      fail: struct
+  type.union:
+    - meta_scope: meta.union.c
+    - include: expression.call
+    - match: ({{identifier}})\s*(?=;)
+      captures:
+        1: entity.name.enum.forward-decl.c
+      pop: 1
+    - match: (?:{{identifier}}\s*)*?({{identifier}})?\s*(?:{{likely_macro}}\s*)*(?=\{|$)
+      captures:
+        1: entity.name.union.c
+      set:
+        - match: \{
+          scope: meta.union.c punctuation.section.block.begin.c
+          set: enclosure.block
+        - match: (?=\S)
+          fail: union
+    - match: (?=\S)
+      fail: union
+
+  enclosure:
+    - match: \(
+      scope: punctuation.section.group.begin
+      push: enclosure.parenthesis
+    - match: \[
+      scope: punctuation.section.brackets.begin
+      push: enclosure.brackets
+    - match: \{
+      scope: punctuation.section.block.begin
+      push: enclosure.block
+  enclosure.parenthesis:
+    - meta_scope: meta.group.c
+    - include: expression
+    - match: \)
+      scope: punctuation.section.group.end
+      pop: 1
+    - match: \}|\] # Fallback
+      scope: invalid.illegal.stray-bracket-end.c
+      pop: 1
```

Comment:
This may easily cause owning contest to also break or not being terminated correctly due to brackets already being consumed here.

I'd recommend to use lookaheads and consume stray brackets only in top-level context.